cmd/runner/testcase.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/xvzcf/tls-interop-runner/internal/pcap"
+	"github.com/xvzcf/tls-interop-runner/internal/utils"
+)
+
+func waitWithTimeout(cmd *exec.Cmd, timeout time.Duration) error {
+	done := make(chan error)
+	go func() {
+		done <- cmd.Wait()
+	}()
+	select {
+	case <-time.After(timeout):
+		if err := cmd.Process.Kill(); err != nil {
+			return fmt.Errorf("failed to kill process:%s", err)
+		}
+		return errors.New("process timed out")
+	case err := <-done:
+		if err != nil {
+			close(done)
+			return err
+		}
+		return nil
+	}
+}
+
+type testCase interface {
+	setup() error
+	run(client endpoint, server endpoint) error
+	verify() error
+}
+
+type testError struct {
+	err         string
+	funcName    string
+	unsupported bool
+}
+
+func (e *testError) Error() string {
+	if e.unsupported {
+		return fmt.Sprintf("Unsupported,%s(): %s", e.funcName, e.err)
+	}
+	return fmt.Sprintf("Failure,%s(): %s", e.funcName, e.err)
+}
+
+type testCaseDC struct {
+	name      string
+	timeout   time.Duration
+	outputDir string
+}
+
+func (t *testCaseDC) setup() error {
+	err := os.MkdirAll(testInputsDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = os.RemoveAll(testOutputsDir)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(testOutputsDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var inputParams strings.Builder
+
+	rootSignatureAlgorithm := utils.SignatureECDSAWithP521AndSHA512
+	err = utils.MakeRootCertificate(
+		&utils.Config{
+			Hostnames:          []string{"root.com"},
+			ValidFrom:          time.Now(),
+			ValidFor:           365 * 25 * time.Hour,
+			SignatureAlgorithm: rootSignatureAlgorithm,
+		},
+		filepath.Join(testInputsDir, "root.crt"),
+		filepath.Join(testInputsDir, "root.key"),
+	)
+	if err != nil {
+		return err
+	}
+	inputParams.WriteString(fmt.Sprintf("Root certificate algorithm: 0x%X\n", rootSignatureAlgorithm))
+
+	intermediateSignatureAlgorithm := utils.SignatureECDSAWithP256AndSHA256
+	err = utils.MakeIntermediateCertificate(
+		&utils.Config{
+			Hostnames:          []string{"example.com"},
+			ValidFrom:          time.Now(),
+			ValidFor:           365 * 25 * time.Hour,
+			SignatureAlgorithm: intermediateSignatureAlgorithm,
+			ForDC:              true,
+		},
+		filepath.Join(testInputsDir, "root.crt"),
+		filepath.Join(testInputsDir, "root.key"),
+		filepath.Join(testInputsDir, "example.crt"),
+		filepath.Join(testInputsDir, "example.key"),
+	)
+	if err != nil {
+		return err
+	}
+	inputParams.WriteString(fmt.Sprintf("Intermediate certificate algorithm: 0x%X\n", intermediateSignatureAlgorithm))
+
+	dcAlgorithm := utils.SignatureECDSAWithP256AndSHA256
```

Comment:
Mmm.. will we be using one type of algorithm?

Reply:
I intended to randomize all these algorithm choices, I just didn't do it yet to make the PR easier to go through, since it's already big.

Reply:
Yeah, this can be a future change 👍